src/test/java/com/commercetools/sync/sdk2/producttypes/ProductTypeSyncTest.java

```diff
@@ -0,0 +1,821 @@
+import com.commercetools.api.models.product_type.ProductTypeChangeNameActionBuilder;
+import com.commercetools.api.models.product_type.ProductTypeDraft;
+import com.commercetools.api.models.product_type.ProductTypeDraftBuilder;
+import com.commercetools.api.models.product_type.ProductTypePagedQueryResponseBuilder;
+import com.commercetools.api.models.product_type.ProductTypeReference;
+import com.commercetools.api.models.product_type.ProductTypeUpdateAction;
+import com.commercetools.sync.sdk2.commons.exceptions.SyncException;
+import com.commercetools.sync.sdk2.producttypes.helpers.ProductTypeSyncStatistics;
+import com.commercetools.sync.sdk2.services.ProductTypeService;
+import com.commercetools.sync.sdk2.services.impl.ProductTypeServiceImpl;
+import io.vrap.rmf.base.client.ApiHttpClient;
+import io.vrap.rmf.base.client.ApiHttpResponse;
+import io.vrap.rmf.base.client.error.BaseException;
+import io.vrap.rmf.base.client.utils.CompletableFutureUtils;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import org.junit.jupiter.api.Test;
+
+class ProductTypeSyncTest {
+
+  @Test
+  void sync_WithEmptyAttributeDefinitions_ShouldSyncCorrectly() {
+    // preparation
+    final ProductTypeDraft newProductTypeDraft =
+        createMockProductTypeDraftBuilder()
+            .key("foo")
+            .name("name")
+            .description("desc")
+            .attributes(emptyList())
+            .build();
+
+    final List<String> errorMessages = new ArrayList<>();
+    final List<Throwable> exceptions = new ArrayList<>();
+
+    final ProductTypeSyncOptions syncOptions =
+        ProductTypeSyncOptionsBuilder.of(mock(ProjectApiRoot.class))
+            .errorCallback(
+                (exception, oldResource, newResource, actions) -> {
+                  errorMessages.add(exception.getMessage());
+                  exceptions.add(exception);
+                })
+            .build();
+
+    final ProductTypeService mockProductTypeService = mock(ProductTypeServiceImpl.class);
+
+    final ProductType existingProductType =
+        createMockProductTypeBuilder().key(newProductTypeDraft.getKey()).build();
+
+    when(mockProductTypeService.fetchMatchingProductTypesByKeys(
+            singleton(newProductTypeDraft.getKey())))
+        .thenReturn(CompletableFuture.completedFuture(singleton(existingProductType)));
+    when(mockProductTypeService.fetchMatchingProductTypesByKeys(emptySet()))
+        .thenReturn(CompletableFuture.completedFuture(Collections.emptySet()));
+    when(mockProductTypeService.updateProductType(any(), any()))
+        .thenReturn(CompletableFuture.completedFuture(existingProductType));
+    when(mockProductTypeService.cacheKeysToIds(anySet()))
+        .thenReturn(CompletableFuture.completedFuture(emptyMap()));
```

Comment:
Looks like this part could be done in a helper method. Can you check if you can use a subclass of ProductTypeServiceImpl and override it's methods?